presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java

```diff
@@ -105,7 +110,7 @@ public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTas
         NodeMap nodeMap = this.nodeMap.get().get();
         NodeAssignmentStats assignmentStats = new NodeAssignmentStats(nodeTaskMap, nodeMap, existingTasks);
 
-        ResettableRandomizedIterator<InternalNode> randomCandidates = randomizedNodes(nodeMap, includeCoordinator, ImmutableSet.of());
+        ResettableRandomizedIterator<InternalNode> randomCandidates = getRandomCandidates(maxTasksPerStage, nodeMap);
```

Comment:
I still don't like the idea with memoizing. The `NodeMap` right now is mutable (although with 5 second memoization). Fixing the selected nodes will change the existing semantics and cluster behaviour. Although i don't know how big a problem is that, i would still prefer to keep behaviour the same.

The `List<RemoteTask> existingTasks` is the list of existing tasks.

`RemoteTask#getNodeId` returns the assigned node id of the task. To keep the existing semantins we should probably change the algorithm to something like:

* Get all the node
* Filter out all the nodes that already have tasks
* Select random N nodes from the remaining nodes, where N  = maxTasksPerStage - existingTasks.size()
* Union nodes with existing tasks ...